### PR TITLE
New Add `vue/no-potential-property-typo` rule

### PR DESCRIPTION
**Pull request by IWANABETHATGUY**
https://github.com/vuejs/eslint-plugin-vue/issues/922

### CODE REVIEW

**Comment by IWANABETHATGUY** [comment]
i make a new rule , `no-potential-typo` rule to warn user a potential typo error, when user define a custom component option is similar to `vue` built-in option, here is an example;
```javascript
export default {
  // ok 
  data() {

  },
  // ok , this maybe a custom property
  testtt: {},
  // bad, this is more likely to be  a typo
  method: {}
}
```

**Comment by ota-meshi** [comment]
Hi @IWANABETHATGUY.
I think it is difficult to judge typos automatically.
I'm worried that this rule may require users to use more suppression comments.

I like a rule to have users set a whitelist and report non-whitelist properties. e.g. `no-unknown-component-options` rule.
Users can also add properties such as `beforeRouteEnter`.

https://router.vuejs.org/guide/advanced/navigation-guards.html#in-component-guards

I like to list similar names in the whitelist as a suggestion to fix the wrong property.

What do you think?

**Comment by IWANABETHATGUY** [comment]
`beforeRouterLeave` is absolute alright for this rule, usually when two string editdistance is. greater than 0 and lessEqual than 1 we report that warning, 
here is the built-in  options  `lib/utils/vue-component-options.json`, we could default this rule for warning not a error.
i don't think people will  define a component option like this
```javascript
export default {
  dat() {},
  method: {},
  comput: {},
  filter: {}
}
```
we default the threshold is 1, only the option in code is very similar but not a vue built-in option will report. we don't check the inner property, so i think this is safe for user;
what do you think?

**Comment by ota-meshi** [comment]
I think people want to detect invalid properties. (The cause of the invalid property may be a typo.)
I think it makes sense to be able to check that properties other than those on the whitelist are not set as a result of the CI.
However, even if you can check that there are no typos (depends on edit distance) as a result of the CI, it doesn't make sense.

I agree that define a component option whitelist is difficult.
We may wish to provide some whitelist presets.
e.g. Vue.js standard, router, vuex, nuxt and more.
Users can whitelist presets and custom property names.

**Comment by IWANABETHATGUY** [comment]
bacuase the indent, the test case is pretty ugly, do you have some good idea for this? and do i need more test case for this rule?

**Comment by ota-meshi** [comment]
Hi @IWANABETHATGUY!
I'm not sure if this rule is best, so I'd like to start with an issue #922.

**Comment by IWANABETHATGUY** [comment]
@ota-meshi could you please help for review

File: lib/rules/no-potential-component-option-typo.js [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+// ------------------------------------------------------------------------------
+// Rule Definition
+// ------------------------------------------------------------------------------
+
+module.exports = {
+  meta: {
+    type: 'suggestion',
+    docs: {
+      description: 'disallow a potential typo in your component property',
+      categories: ['essential'],
```
Comment (ota-meshi) [inline lib/rules/no-potential-component-option-typo.js]: Could you change the categories to be set to undefined?
New rules set categories at major release.

File: tests/lib/rules/no-potential-component-option-typo.js [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+
+      filename: 'test.vue',
+      code: `
+      <script>
+      export default {
+        data() {}
+      }
+      </script>
+      `.trim(),
+      options: [{ presets: ['vue'] }]
```
Comment (ota-meshi) [inline tests/lib/rules/no-potential-component-option-typo.js]: Could you add test cases using a preset other than vue and test cases using multiple presets.

File: tests/lib/rules/no-potential-component-option-typo.js [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+        dat: {},
+        methods: {}
+      }
+      </script>
+              `.trim()
+            }
+          ]
+        }
+      ],
+      options: [{ custom: ['data', 'methods'] }]
```
Comment (ota-meshi) [inline tests/lib/rules/no-potential-component-option-typo.js]: Could you use custom option test cases that is not available in the presets?

File: tests/lib/rules/no-potential-component-option-typo.js [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+      filename: 'test.vue',
+      code: `
+      <script>
+      export default {
+        data() {}
+      }
+      </script>
+      `.trim(),
+      options: [{ presets: ['vue'] }]
+    }
```
Comment (ota-meshi) [inline tests/lib/rules/no-potential-component-option-typo.js]: Could you use test cases that are valid using custom and threshold?

File: lib/rules/no-potential-component-option-typo.js [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+          }
+        }
+      }
+    ]
+  },
+
+  create: function (context) {
+    const option = context.options[0] || {}
+    const custom = option['custom'] || []
+    const presets = option['presets'] || []
```
Comment (ota-meshi) [inline lib/rules/no-potential-component-option-typo.js]: Could you change Vue.js standard options to validate by default.

File: tests/lib/rules/no-potential-component-option-typo.js [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+    {
+      filename: 'test.vue',
+      code: `
+      <script>
+      export default {
+        dat: {},
+        method: {}
+      }
+      </script>
+      `.trim(),
```
Comment (ota-meshi) [inline tests/lib/rules/no-potential-component-option-typo.js]: Could you remove `trim()`?

Comment (ota-meshi) [reply tests/lib/rules/no-potential-component-option-typo.js]: Please change other same points.

File: tests/lib/rules/no-potential-component-option-typo.js [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+      errors: [
+        {
+          message: "'dat' may be a typo, which is similar to option [data].",
```
Comment (ota-meshi) [inline tests/lib/rules/no-potential-component-option-typo.js]: Could you test the location of the error using the `line` and `column` properties?

File: tests/lib/rules/no-potential-component-option-typo.js [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+      options: [{ presets: ['vue'] }]
+    },
+    // test if give preset and the potentialTypoList length is zero, just for 100% test cover
+    {
+
+      filename: 'test.vue',
+      code: `
+      <script>
+      export default {
+        data() {}
```
Comment (ota-meshi) [inline tests/lib/rules/no-potential-component-option-typo.js]: Could you add a test case that tests all the names of the presets?

File: tests/lib/rules/no-potential-component-option-typo.js [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+      }
+      </script>
+      `.trim(),
+      options: [{ presets: ['vue'] }]
+    }
+    // give me some code that won't trigger a warning
```
Comment (ota-meshi) [inline tests/lib/rules/no-potential-component-option-typo.js]: Could you add test case that set `custom: ['abcde', 'abcd']` as option and vue with options for `'abcde'` and `'abcd'`?

File: lib/rules/no-potential-component-option-typo.js [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+    const threshold = option['threshold'] || 1
+    let candidateOptions = []
```
Comment (ota-meshi) [inline lib/rules/no-potential-component-option-typo.js]: The assigned value does not seem to be used. Remove the assignment or change it to use `push(...arg)`.

File: lib/rules/no-potential-component-option-typo.js [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+    if (presets.includes('all')) {
+      candidateOptions = Object.keys(vueComponentOptions).reduce((pre, cur) => {
+        return [...pre, ...vueComponentOptions[cur]]
+      }, [])
+    } else {
+      candidateOptions = presets.reduce((pre, cur) => {
+        return [...pre, ...vueComponentOptions[cur]]
+      }, [])
+    }
+    candidateOptions = [...new Set(candidateOptions.concat(custom))]
```
Comment (ota-meshi) [inline lib/rules/no-potential-component-option-typo.js]: Destructuring can be used.

```suggestion
    candidateOptions = [...new Set([...candidateOptions,...custom])]
```

File: tests/lib/rules/no-potential-component-option-typo.js (hunk 8 above)
Comment (IWANABETHATGUY) [reply tests/lib/rules/no-potential-component-option-typo.js:55]: could you please make it more clear, i was wonder if you mean all the options in preset.json, or add a test case, that presets option is ['vue', 'vue-router', 'nuxt']

**Comment by IWANABETHATGUY** [comment]
i still have a change request, but i have some question, could please give me some tips?

File: tests/lib/rules/no-potential-component-option-typo.js (hunk 8 above)
Comment (ota-meshi) [reply tests/lib/rules/no-potential-component-option-typo.js:56]: I want you to add a test case that can verify that all of the names defined in `vue-component-options.json` are valid.

**Comment by ota-meshi** [comment]
I want you to add a test case that can verify that all of the names defined in `vue-component-options.json` are valid.

File: tests/lib/rules/no-potential-component-option-typo.js (hunk 8 above)
Comment (IWANABETHATGUY) [reply tests/lib/rules/no-potential-component-option-typo.js:65]: finish that😊

File: tests/lib/rules/no-potential-component-option-typo.js [hunk 12]
```diff
@@ -0,0 +1,406 @@
...
+        fooooo: {}
+      }
+      </script>
+      `,
+      options: [{ custom: ['custom', 'foo'], threshold: 2 }]
+    },
+    // test all valid vue options
+    {
+      filename: 'test.vue',
+      code: VueComponentJsonCheckVue,
```
Comment (ota-meshi) [inline tests/lib/rules/no-potential-component-option-typo.js:125]: Hmm.. I don't think it makes sense to put it in a separate file.
Could you put the code inline here?

**Pull request merged**